charlesm93/math
# Issue 16 autodiff tests

#### Submisison Checklist

- [X] Run unit tests: `./runTests.py test/unit/torsten`

#### Summary:
Resolves parts of issue #16. The progress is significant enough to merge this branch `torsten-develop`.
 
Tests automatic differentiation (auto-diff) for parameters stored in `pMatrix`, the parameter matrix and `system`, the rate constant matrix. Creates `util` files with functions that computed differentiation and compare the result to auto-diff. 

#### Intended Effect:
Unit tests test computation of the gradient, with respect to `pMatrix` and `system`.

#### How to Verify:
Finite-diff and auto-diff are in close agreement (1e-4, with some notable and expected exceptions in functions that use numerical integrators). A third independent check would be desirable, though I'm not sure how to get one.

#### Side Effects:
None.

#### Documentation:
None. I hope the code is readable. I closely followed the example of `integrate_ode_bdf`. The finite differentiation functions are straightforward. Most of the difficulty comes from the heavy bookkeeping. Each function returns a matrix and takes in a lot of arguments (i.e. a lot of dependent variables to derive w.r.t a lot of independent variables). 

#### Further Notes: 
I'll need to extend the tests to handle parameters, which are not stored in `pMatrix` and `system`. And `tlags`, once issue #17 is resolved.


#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Metrum Research Group LLC

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
